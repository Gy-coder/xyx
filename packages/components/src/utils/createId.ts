let id = 0;

export default function createId(){
    return id++
}