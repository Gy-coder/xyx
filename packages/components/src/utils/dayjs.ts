type DayUnit = "year"
    | "month"
    | "day"
    | "hour"
    | "minute"
    | "second"
    | "millisecond";

export default class Dayjs {
    public date: Date

    constructor(date?: number | string | Date) {
        this.date = date ? new Date(date) : new Date()
    }

    get year() {
        return this.date.getFullYear()
    }

    get month() {
        return this.date.getMonth() + 1
    }

    get week() {
        return this.date.getDay()
    }

    get day() {
        return this.date.getDate()
    }

    get hour() {
        return this.date.getHours()
    }

    get minute() {
        return this.date.getMinutes()
    }

    get second() {
        return this.date.getSeconds()
    }

    get millisecond() {
        return this.date.getMilliseconds()
    }

    get raw() {
        return this.date
    }

    public firstDayOfMonth() {
        return new Dayjs(
            new Date(
                this.year,
                this.month - 1,
                1,
                0,
                0,
                0)
        )
    }

    public add(amount: number, unit: DayUnit) {
        let date = new Date(this.date.getTime());
        switch (unit) {
            case 'year':
                const currentDate = date.getDate()
                date.setDate(1)
                date.setFullYear(date.getFullYear() + amount)
                const targetDate = new Date(
                    date.getFullYear(),
                    date.getMonth() + 1,
                    0,
                    0,
                    0,
                    0,
                ).getDate()
                date.setDate(Math.min(currentDate, targetDate))
                break;
            case 'month':
                const d = date.getDate()
                date.setDate(1)
                date.setMonth(date.getMonth() + amount);
                const d2 = new Date(date.getFullYear(), date.getMonth() + 1, 0, 0, 0, 0).getDate()
                date.setDate(Math.min(d, d2))
                break;
            case 'day':
                date.setDate(date.getDate() + amount);
                break;
            case 'hour':
                date.setHours(date.getHours() + amount);
                break;
            case 'minute':
                date.setMinutes(date.getMinutes() + amount);
                break;
            case 'second':
                date.setSeconds(date.getSeconds() + amount);
                break;
            case 'millisecond':
                date.setMilliseconds(date.getMilliseconds() + amount);
                break;

            default:
                throw new Error('Time.add: unknown unit');
        }
        return new Dayjs(date)
    }

    public set(amount: number,unit: DayUnit){
        const date = new Date(this.date.getTime())
        switch (unit){
            case "year":
                date.setFullYear(amount)
                break
            case "month":
                date.setMonth(amount - 1)
                break
            case "day":
                date.setDate(amount)
                break
            case "hour":
                date.setHours(amount)
                break
            case "second":
                date.setSeconds(amount)
                break
            case "millisecond":
                date.setMilliseconds(amount)
                break
            default:
                throw new Error("Time.add: unknown unit")
        }
        return new Dayjs(date)
    }

    isSameDay(day: Dayjs){
        return day.year === this.year && day.month === this.month && day.day === this.day
    }

    isSameMonth(day: Dayjs){
        return day.year === this.year && day.month === this.month
    }

    //@ts-ignore
    format(pattern?: string = 'YYYY-MM-DD') {
        // 目前支持的格式有 YYYY MM DD HH mm ss SSS
        return pattern.replace(/YYYY/g, this.year.toString())
            .replace(/MM/, this.month.toString().padStart(2, '0'))
            .replace(/DD/, this.day.toString().padStart(2, '0'))
            .replace(/HH/, this.hour.toString().padStart(2, '0'))
            .replace(/mm/, this.minute.toString().padStart(2, '0'))
            .replace(/ss/, this.second.toString().padStart(2, '0'))
            .replace(/SSS/, this.millisecond.toString().padStart(3, '0'))
    }
}
