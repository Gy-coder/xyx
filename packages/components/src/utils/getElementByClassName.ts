export function getElementByClassName(container = document, classname: string) {
  return container.getElementsByClassName(classname)[0];
}
