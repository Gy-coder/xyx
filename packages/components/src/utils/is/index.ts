export { default as isUndefined } from "./isUndefined";
export { default as isFunction } from "./isFunction";
