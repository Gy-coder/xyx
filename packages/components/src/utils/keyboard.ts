export const keyboard = {
  Enter: "Enter",
};
