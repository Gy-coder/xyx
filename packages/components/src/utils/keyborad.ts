export const keyborad = {
  Enter: "Enter",
};
