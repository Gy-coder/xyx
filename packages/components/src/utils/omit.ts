export default function omit<T = any>(obj: Record<string, T>, paths: string[]) {
  const res: Record<string, T> = { ...obj },
    set = new Set(paths);
  for (let key in obj) {
    if (set.has(key)) delete res[key];
  }
  return res;
}
