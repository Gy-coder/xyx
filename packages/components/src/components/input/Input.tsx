import {
  forwardRef,
  ForwardRefRenderFunction,
  ChangeEventHandler,
  MouseEventHandler,
  KeyboardEventHandler,
} from "react";
import classnames from "classnames";
import useMergeState from "../../hooks/useMergeState";
import InternalIcon from "../_interal/internal_icon/InternalIcon";
import { InputProps } from "./interface";
import useHoverFocus from "./useHoverFocus";
import { keyboard } from "../../utils/keyboard";
import "./index.scss";

const Input: ForwardRefRenderFunction<any, InputProps> = (props, ref) => {
  const {
    defaultValue,
    value,
    onChange,
    size = "middle",
    clearable = false,
    onFocus,
    onBlur,
    onMouseEnter,
    onMouseLeave,
    onPressEnter,
    prefix,
    suffix,
    maxLength,
    showCount = false,
    bordered = true,
    disabled = false,
    status,
    rounded = false,
    ...rest
  } = props;

  const [stateValue, setStateValue, isControlled] = useMergeState("", {
    defaultValue,
    value,
  });
  const {
    hover,
    focus,
    handleBlur,
    handleFocus,
    handleMouseEnter,
    handleMouseLeave,
  } = useHoverFocus({ onBlur, onFocus, disabled });
  const classes = classnames("g-input", {
    [`g-input-focus`]: focus,
    [`g-input-hover`]: hover,
    [`g-input-${size}`]: size,
    [`g-input-borderless`]: !bordered,
    [`g-input-disabled`]: disabled,
    [`g-input-${status}`]: status,
    [`g-input-rounded`]: rounded,
  });

  const handleChange: ChangeEventHandler<HTMLInputElement> = (e) => {
    if (disabled) return;
    if (!isControlled) {
      setStateValue(e.target.value);
    }
    onChange?.(e.target.value, e);
  };
  const handleReset: MouseEventHandler = () => {
    if (!isControlled) {
      setStateValue("");
    }
    onChange?.("");
  };
  const handleKeyDown: KeyboardEventHandler = (e) => {
    if (e.code === keyboard.Enter) onPressEnter?.(e);
  };
  const prefixIcon = prefix ? (
    <span className="g-input-prefix">{prefix}</span>
  ) : null;
  const suffixIcon = suffix ? (
    <span className="g-input-suffix">{suffix}</span>
  ) : null;
  const renderShowCount = showCount ? (
    <span className="g-input-show-count">
      {stateValue.length}
      {maxLength && <>&nbsp;/&nbsp;{maxLength}</>}
    </span>
  ) : null;
  const clearButton =
    clearable && stateValue !== "" && !disabled ? (
      <span className="g-input-clear" onClick={handleReset}>
        <InternalIcon />
      </span>
    ) : null;
  return (
    <div
      className={classes}
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
    >
      {prefixIcon}
      <input
        ref={ref}
        type="text"
        className={classnames("g-input-origin")}
        value={stateValue}
        onChange={handleChange}
        onFocus={handleFocus}
        onBlur={handleBlur}
        maxLength={maxLength}
        onKeyDown={handleKeyDown}
        disabled={disabled}
        {...rest}
      />
      {clearButton}
      {renderShowCount}
      {suffixIcon}
    </div>
  );
};

export default forwardRef(Input);
