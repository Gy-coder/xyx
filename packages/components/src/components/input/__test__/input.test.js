import { screen, render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import Input from "../";
import { getElementByClassName } from "../../../utils/getElementByClassName";
import { keyboard } from "../../../utils/keyboard";
import { useState } from "react";

describe("test Input component", () => {
  it("can render correctly", () => {
    render(<Input placeholder="input" />);
    expect(screen.getByPlaceholderText("input")).toBeVisible();
  });
  it("snapshot", () => {
    const { asFragment } = render(<Input />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("can change value", () => {
    const jestFn = jest.fn();
    render(<Input defaultValue="hello" onChange={jestFn} />);
    const el = screen.getByDisplayValue("hello");
    fireEvent.change(el, { target: { value: "i love u" } });
    expect(jestFn).toHaveBeenCalledTimes(1);
    expect(el.value).toBe("i love u");
  });
  it("size middle", () => {
    const { container } = render(<Input placeholder="123" />);
    const el = container.getElementsByClassName("g-input")[0];
    expect(el).toHaveClass("g-input-middle");
  });
  it("size large", () => {
    const { container } = render(<Input placeholder="123" size="large" />);
    const el = container.getElementsByClassName("g-input")[0];
    expect(el).toHaveClass("g-input-large");
  });
  it("size small", () => {
    const { container } = render(<Input placeholder="123" size="small" />);
    const el = container.getElementsByClassName("g-input")[0];
    expect(el).toHaveClass("g-input-small");
  });
  it("can clear value", () => {
    const { container } = render(<Input placeholder="123" clearable />);
    const el = container.getElementsByClassName("g-input-clear")[0];
    const input = screen.getByPlaceholderText("123");
    expect(el).toBeFalsy();
    fireEvent.change(input, { target: { value: "ccccc" } });
    const el2 = container.getElementsByClassName("g-input-clear")[0];
    expect(el2).toBeVisible();
    expect(input).toHaveValue("ccccc");
    fireEvent.click(el2);
    const input2 = container.getElementsByClassName("g-input-origin")[0];
    expect(input2).toHaveValue("");
  });
  it("can render prefix", () => {
    const { container, asFragment } = render(<Input prefix={<div>123</div>} />);
    const el = container.getElementsByClassName("g-input-prefix")[0];
    expect(el).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });
  it("can render suffix", () => {
    const { container, asFragment } = render(<Input suffix={<div>123</div>} />);
    const el = container.getElementsByClassName("g-input-suffix")[0];
    expect(el).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });
  it("can render showCount and redender correct", () => {
    const { container, asFragment } = render(
      <Input defaultValue="123" showCount />
    );
    const el = container.getElementsByClassName("g-input-show-count")[0];
    expect(el).toBeVisible();
    expect(el.textContent).toBe("3");
    expect(asFragment()).toMatchSnapshot();
  });
  it("can render showCount and redender correct that has maxLength property", () => {
    const { container, asFragment } = render(
      <Input defaultValue="123" showCount maxLength={20} />
    );
    const el = container.getElementsByClassName("g-input-show-count")[0];
    expect(el).toBeVisible();
    expect(el.innerHTML).toBe("3&nbsp;/&nbsp;20");
    expect(asFragment()).toMatchSnapshot();
  });
  it("can add border false", () => {
    const { container } = render(<Input defaultValue="123" bordered={false} />);
    const el = getElementByClassName(container, "g-input");
    expect(el).toHaveClass("g-input-borderless");
  });
  it("can have error status", () => {
    const { container } = render(<Input status="error" placeholder="error" />);
    const el = getElementByClassName(container, "g-input");
    expect(el).toBeVisible();
    expect(el).toHaveClass("g-input-error");
  });
  it("can have warning status", () => {
    const { container } = render(
      <Input status="warning" placeholder="warning" />
    );
    const el = getElementByClassName(container, "g-input");
    expect(el).toBeVisible();
    expect(el).toHaveClass("g-input-warning");
  });
  it("can rounded", () => {
    const { container } = render(<Input placeholder="1234" rounded />);
    const el = getElementByClassName(container, "g-input");
    expect(el).toHaveClass("g-input-rounded");
  });
  it("disabled", () => {
    const jestFn = jest.fn();
    const { container } = render(
      <Input defaultValue="123" onChange={jestFn} disabled clearable />
    );
    const el = getElementByClassName(container, "g-input");
    const input = screen.getByDisplayValue("123");
    expect(el).toHaveClass("g-input-disabled");
    fireEvent.change(input, { target: { value: "ccc" } });
    expect(jestFn).not.toBeCalled();
    expect(input).toHaveValue("123");
    const clearButton = getElementByClassName(container, "g-input-clear");
    expect(clearButton).toBeFalsy();
  });
  it("pressEnter", () => {
    const jestFn = jest.fn();
    render(<Input placeholder="123" onPressEnter={jestFn} />);
    const el = screen.getByPlaceholderText("123");
    fireEvent.keyDown(el, { code: keyboard.Enter });
    expect(jestFn).toBeCalledTimes(1);
    fireEvent.keyDown(el, { code: keyboard.Enter });
    expect(jestFn).toBeCalledTimes(2);
  });
  it("disabled pressEnter", () => {
    const jestFn = jest.fn();
    render(<Input placeholder="123" onPressEnter={jestFn} disabled />);
    const el = screen.getByPlaceholderText("123");
    fireEvent.keyDown(el, { code: keyboard.Enter });
    expect(jestFn).not.toBeCalled();
    fireEvent.keyDown(el, { code: keyboard.Enter });
    expect(jestFn).not.toBeCalled();
  });
  it("controlled component", () => {
    const jestFn = jest.fn();
    let v, setV, handleChange;
    const Demo = () => {
      [v, setV] = useState("");
      handleChange = (newValue) => {
        setV(newValue);
        jestFn(newValue);
      };
      return <Input value={v} onChange={handleChange} placeholder="input" />;
    };
    render(<Demo />);
    const el = screen.getByPlaceholderText("input");
    fireEvent.change(el, { target: { value: "ccccc" } });
    expect(v).toBe("ccccc");
    expect(jestFn).toBeCalledTimes(1);
    expect(jestFn).toBeCalledWith("ccccc");
  });
  it("addOnBefore", () => {
    const { contaier } = render(
      <Input addOnBefore={<>123</>} placeholder="123" />
    );
    const el = getElementByClassName(contaier, "g-input-add-on-before");
    const input = screen.getByPlaceholderText("123");
    expect(el).toBeVisible();
    expect(input).toBeVisible();
  });
  it("addOnAfter", () => {
    const { contaier } = render(
      <Input addOnAfter={<>123</>} placeholder="123" />
    );
    const el = getElementByClassName(contaier, "g-input-add-on-after");
    const input = screen.getByPlaceholderText("123");
    expect(el).toBeVisible();
    expect(input).toBeVisible();
  });
});

describe("test Input password", () => {
  it("test password", () => {
    const { container } = render(<Input.Password />);
    const el = getElementByClassName(container, "g-input-origin");
    const visibleIcon = getElementByClassName(
      container,
      "g-input-visibility-icon"
    );
    expect(el.type).toBe("password");
    fireEvent.click(visibleIcon);
    expect(el.type).toBe("text");
    fireEvent.click(visibleIcon);
    expect(el.type).toBe("password");
  });
  it("visibility false", () => {
    const { container } = render(<Input.Password visibility={false} />);
    const el = getElementByClassName(container, "g-input-visibility-icon");
    expect(el).not.toBeTruthy();
  });
});
