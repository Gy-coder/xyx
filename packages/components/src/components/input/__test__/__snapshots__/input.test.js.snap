// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Input component can render prefix 1`] = `
<DocumentFragment>
  <div
    class="g-input g-input-middle"
  >
    <span
      class="g-input-prefix"
    >
      <div>
        123
      </div>
    </span>
    <input
      class="g-input-origin"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`test Input component can render showCount and redender correct 1`] = `
<DocumentFragment>
  <div
    class="g-input g-input-middle"
  >
    <input
      class="g-input-origin"
      type="text"
      value="123"
    />
    <span
      class="g-input-show-count"
    >
      3
    </span>
  </div>
</DocumentFragment>
`;

exports[`test Input component can render showCount and redender correct that has maxLength property 1`] = `
<DocumentFragment>
  <div
    class="g-input g-input-middle"
  >
    <input
      class="g-input-origin"
      maxlength="20"
      type="text"
      value="123"
    />
    <span
      class="g-input-show-count"
    >
      3 / 20
    </span>
  </div>
</DocumentFragment>
`;

exports[`test Input component can render suffix 1`] = `
<DocumentFragment>
  <div
    class="g-input g-input-middle"
  >
    <input
      class="g-input-origin"
      type="text"
      value=""
    />
    <span
      class="g-input-suffix"
    >
      <div>
        123
      </div>
    </span>
  </div>
</DocumentFragment>
`;

exports[`test Input component snapshot 1`] = `
<DocumentFragment>
  <div
    class="g-input g-input-middle"
  >
    <input
      class="g-input-origin"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;
